csharp/extractor/Semmle.Extraction.CSharp/Entities/PreprocessorDirectives/PragmaWarningDirective.cs

```diff
@@ -0,0 +1,43 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Semmle.Extraction.Entities;
+using System.IO;
+
+namespace Semmle.Extraction.CSharp.Entities
+{
+    internal class PragmaWarningDirective : FreshEntity, IPreprocessorDirective
+    {
+        private readonly PragmaWarningDirectiveTriviaSyntax trivia;
+
+        public PragmaWarningDirective(Context cx, PragmaWarningDirectiveTriviaSyntax trivia)
+            : base(cx)
+        {
+            this.trivia = trivia;
+            TryPopulate();
+        }
+
+        protected override void Populate(TextWriter trapFile)
+        {
+            trapFile.pragma_warnings(this, trivia.DisableOrRestoreKeyword.IsKind(SyntaxKind.DisableKeyword) ? 0 : 1);
+
+            var childIndex = 0;
+            foreach (var code in trivia.ErrorCodes)
+            {
+                trapFile.pragma_warning_error_codes(this, code.ToString(), childIndex++);
+            }
+
+            trapFile.preprocessor_directive_location(this, cx.Create(ReportingLocation));
+
+            if (!cx.Extractor.Standalone)
+            {
+                var assembly = Assembly.CreateOutputAssembly(cx);
+                trapFile.preprocessor_directive_assembly(this, assembly);
```

Comment:
Would it be better to map to the compilation instead? I realize that the output assembly will, in most cases, determine the compilation, but we might have two compilations that output the same assembly.

Reply:
We could do that, but I think we would need to change the `Compilation` to be a cached entity then. Should I do it?

Reply:
Or perhaps just store it in a field on the `Context`?

Reply:
I thought this would work, but adding the compilation entity to the context is also not straightforward for two reasons:
* syntax tree extraction and compilation extraction have different `context` instances,
* extraction happens on multiple threads, so we can't be sure that compilation is extracted before directives.